File: 0056. Merge Intervals.py
"""
Method 1:
Sort the intervals first and then check if the intervals need to be combined.

Time complexity:
Space complexity: O(1)
"""

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        def combine(lst1, lst2):
            return [min(lst1[0], lst2[0]), max(lst1[1], lst2[1])]
        
        intervals = sorted(intervals, key=lambda x: x[0])
        i = 0
        while i < len(intervals)-1:
            if not intervals[i+1][0] > intervals[i][1]:
                newInterval = combine(intervals[i], intervals[i+1])
                intervals[i] = newInterval
                intervals.pop(i+1)
            else:
                i += 1
        return intervals

File: 0056. Merge Intervals.cpp
/*
Method 1:
Sort the intervals by first element in vector in intervals.
Then, if 2 intervals intersect, merge them. If not, add the interval into result.

Time complexity: O(nlogn)
Space complexity: O(N)
*/

class Solution {
public:
    static bool sortcol(const vector<int>& v1, const vector<int>& v2){
            return v1[0] < v2[0];
        }

    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), sortcol);
        vector<vector<int>> result;

        for (int i = 1; i < intervals.size(); i++){
            if (intervals[i-1][1] >= intervals[i][0])
            {
                vector<int> newInterval = {intervals[i-1][0], max(intervals[i-1][1], intervals[i][1])};
                intervals[i] = newInterval;
            }
            else {
                result.push_back(intervals[i-1]);
            }
        }
        result.push_back(intervals.back());
        return result;
    }
};
